pkg/distributor/ratestore.go

```diff
@@ -149,29 +167,24 @@ func (s *rateStore) anyShardingEnabled() bool {
 	return false
 }
 
-func (s *rateStore) aggregateByShard(streamRates map[string]*logproto.StreamRate) map[string]int64 {
-	shardCount := make(map[string]int)
-	rates := make(map[string]int64)
-
-	for _, sr := range streamRates {
-		key := key(sr.Tenant, sr.StreamHashNoShard)
-		shardCount[key]++
+func (s *rateStore) aggregateByShard(streamRates map[string]map[uint64]*logproto.StreamRate) map[string]map[uint64]expiringRate {
+	rates := map[string]map[uint64]expiringRate{}
 
-		if _, ok := rates[key]; ok {
-			rates[key] += sr.Rate
-			continue
-		}
+	for tID, tenant := range streamRates {
+		for _, streamRate := range tenant {
+			if _, ok := rates[tID]; !ok {
+				rates[tID] = map[uint64]expiringRate{}
+			}
 
-		rates[key] = sr.Rate
-	}
+			rate := rates[tID][streamRate.StreamHashNoShard]
+			rate.rate += streamRate.Rate
+			rate.shards++
+			rate.createdAt = time.Now()
```

Comment:
`createdAt` is probably better named as `updatedAt` the idea is that when a stream hasn't showed up for awhile, we expire it. Otherwise we just keep updating it.

Reply:
ohh got it, aggregateByShard is only invoked during updates. resolved!